### PR TITLE
feat: revamp production and testing docker + update to node 20 + revamp the Node targets

### PR DESCRIPTION
**Pull request by abeimler**
Hi,
I try to update node from [12 -> 16](https://nodejs.dev/en/about/releases/) and add Node 18.
_Not 1000% sure how the build system works with pnpm and `node12`-`node16`, did some changes in the [package.json](package.json)..._

_My goal was first to build and test setup-cpp with different node versions and then dump the node version._
_So every contributor is working with the same version or higher (may via nvm)._

EDIT:
We also can build (and package) with node 16+ and then test( in docker) for different node versions (12-18).

## Update

* Add (CI) Tests for building/testing with Node 12, 16 and 18
* Update docker files and use [nvm](https://github.com/nvm-sh/nvm) for installing Node version
* update/specify [pnpm version](https://pnpm.io/installation#compatibility) for node version 
* run `pnpm audit --fix`

_Node 12 is no longer maintained and Node 16 is ending in [few months](https://endoflife.date/nodejs)_

### Environment

**OS:** Windows 10
**Node:** 18.15.0
**pnpm:** 7.4.0

### Tests with Node 12

I added Tests for node 12, 16 and 18.
But it seems like, node 12 didn't work:
```bash
/home/runner/work/setup-cpp/setup-cpp/node_modules/.pnpm/jest-cli@29.3.1_c94e0f8eadb71b347b2a0c987637d777/node_modules/jest-cli/build/run.js:129
    if (error?.stack) {
              ^

SyntaxError: Unexpected token '.'
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
 ELIFECYCLE  Command failed with exit code 1.
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/home/runner/work/setup-cpp/setup-cpp/node_modules/.pnpm/jest-cli@29.3.1_c94e0f8eadb71b347b2a0c987637d777/node_modules/jest-cli/build/index.js:12:12)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
/home/runner/work/setup-cpp/setup-cpp/node_modules/.pnpm/cspell@6.19.2/node_modules/cspell/dist/app.js:26
        .argParser((value, prev) => prev?.concat(value) || [value]);
                                         ^

SyntaxError: Unexpected token '.'
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/home/runner/work/setup-cpp/setup-cpp/node_modules/.pnpm/cspell@6.19.2/node_modules/cspell/bin.js:4:13)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
```
https://github.com/abeimler/setup-cpp/actions/runs/4631005113/jobs/8193334766

EDIT: I also try to skip the Test and run "Create Executable" with Node 12, got some errors...
```bash
(node:2914) UnhandledPromiseRejectionWarning: Error: Command failed with exit code 1: /opt/hostedtoolcache/node/12.22.12/x64/bin/node --require /home/runner/work/setup-cpp/setup-cpp/node_modules/.pnpm/ts-node@10.9.1_0581ab4a3d99d86da3e98604f38a692c/node_modules/ts-node/dist/child/child-require.js --loader file:///home/runner/work/setup-cpp/setup-cpp/node_modules/.pnpm/ts-node@10.9.1_0581ab4a3d99d86da3e98604f38a692c/node_modules/ts-node/child-loader.mjs /home/runner/work/setup-cpp/setup-cpp/node_modules/.pnpm/ts-node@10.9.1_0581ab4a3d99d86da3e98604f38a692c/node_modules/ts-node/dist/child/child-entrypoint.js --brotli-base64-config=G1gEABwHdpPP1bayIuVofKpFtmD2TrZlr/9+KQTRczTguzVHpSRImblm0enqO3UTWL89aC2gvYW6zhN6IAs19tTproEiHmo3Orsm8fSpfuGbdyjHK1EmEFwpeKBGztAX6119SncaRuGY5QsHNhWYtAbmZRjyGUzKARTeyXSanlCY8R+vB0OAmoapBxiRx2v/MTSlUMwHLchMnPAxlylKJotZzOWKLNJ8wcmxqLCpG9JBAmQ1wiAv+B1HLjsWe4Jq0AnoR+pcrhUR4XexBZzZgHHC4Ec6KGywpHPAs8ewFHsBewPsIg+Seuj5tJN46ySZzCPuahItlsiTKP+ZfkOm1VVts4+5nefOuSxQXHFaPhyhZEXPD2sxBdoyhnV2MN71atAXjzNofM8RK3Ilg0iiRpu129u0uSn4zGlF5ja6To16f759wjNbeFWsZFXRe1PZmsaTY0Pcn4Omuny5X9KKTQT9LtjVYvAG49bA0sLLJce519DC0vw0CQNxW167hKhp0MyFDYZxMJxjZdFhdJhyHUjfEhdMt3XveeqShgeB/eOmSZwyK8ejd6zhxqGfDhddP3R2tSSEcP8E6IrTUty4DyUOkRymaa0H29ZcHsaIVXnewLy+bkcAb0Sx1yPl+7iGy8PT/aa7tlTytFpxd3r8uO2QMZ0t46J5D4HR3UAnfdQuhMV0X2z1/H1vunfZuFWFLj5rD4SxttARVyXd23UIIummMNvKemTU7fdvVyiQ5Sv4NrBAbWBkvucURqehL746lwibGwajDKfoSLAf2NRdIVtAJUqh1GhMQag2FhXKbuWop7n5Hy0+sUS5JN0yghunHw== ./node_modules/caxa/build/index.mjs --input ./dist/node16 --output ./exe/setup-cpp-x64-linux -- {{caxa}}/node_modules/.bin/node {{caxa}}/setup-cpp.js
SyntaxError: Unexpected reserved word
    at Loader.moduleStrategy (internal/modules/esm/translators.js:140:18)
    at async link (internal/modules/esm/module_job.js:42:21)
    at makeError (/home/runner/work/setup-cpp/setup-cpp/node_modules/.pnpm/execa@5.1.1/node_modules/execa/lib/error.js:60:[11](https://github.com/abeimler/setup-cpp/actions/runs/4721345345/jobs/8374515219#step:8:12))
    at handlePromise (/home/runner/work/setup-cpp/setup-cpp/node_modules/.pnpm/execa@5.1.1/node_modules/execa/index.js:118:26)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async Promise.all (index 0)
(node:29[14](https://github.com/abeimler/setup-cpp/actions/runs/4721345345/jobs/8374515219#step:8:15)) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)

```
 

https://github.com/abeimler/setup-cpp/actions/runs/4721345345/jobs/8374515219

_For me it seems like the test and package stage didn't work with Node 12 (?)_

### ~~Audit~~

~~run `pnpm audit --fix` to update some dependencies:~~
```json
    "overrides": {
      ...
      "got@<11.8.5": ">=11.8.5",
      "minimist@<0.2.1": ">=0.2.1",
      "http-cache-semantics@<4.1.1": ">=4.1.1",
      "minimist@<0.2.4": ">=0.2.4",
      "deep-extend@<0.5.1": ">=0.5.1"
    },
```

Seems to be fixed with the latest dependencies update.

### get rid of `dist/`-folder

Personally, I would like to get rid of `dist/`-folder and just build everything with the CI, then release the right version via CI. _(best would be an release via [npm package](https://github.com/aminya/setup-cpp/issues/126))_
_This can also help to avoid conflicts in `dist/`_

Then remove or deprecate Node12 in [package.json](package.json) (`main.legacy`, `copy.matchers.node12`, ...), update 
```
  "engines": {
    "node": ">=12.x"
  },
```
and/or make node 16 the new (minimal) requirement.

### Docker

* [x] Update dockerfiles and make them multi-stage with `build`- and `run/test`-stage, so you don't need to copy from `dist/`.

### CODE REVIEW

**Comment by aminya** [comment]
I'll look into this PR soon. But keep in mind that Node 12 and Node 16 are needed for GitHub Actions.

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -23,12 +23,20 @@ jobs:
...
+          - 16
+          - 18
+        include:
+          - node: 12
+            pnpm: 6
+            skip_test: true
+          - node: 16
+            pnpm: 7
+          - node: 18
+            pnpm: 8
```
Comment (aminya) [inline .github/workflows/CI.yml:39]: We don't need to test with different pnpm versions. Please just specify the latest as the version

```suggestion
        pnpm:
          - 8
```

**Review by aminya — CHANGES_REQUESTED** [review]

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -62,7 +64,7 @@ jobs:
       #   uses: mxschmitt/action-tmate@v3
 
       - name: Test
-        if: "${{ !contains(github.event.head_commit.message, '[skip test]') }}"
+        if: "${{ !contains(github.event.head_commit.message, '[skip test]') && !matrix.skip_test }}"
```
Comment (aminya) [inline .github/workflows/CI.yml:67]:
```suggestion
        if: "${{ !contains(github.event.head_commit.message, '[skip test]') && contains(matrix.node, '18') }}"
```

**Review by aminya — CHANGES_REQUESTED** [review]

File: dev/docker/ubuntu_20.04_node.dockerfile [hunk 3]
```diff
@@ -4,36 +4,59 @@ FROM ubuntu:20.04 as base
...
-WORKDIR "/"
 
-# run installation
-RUN node ./setup-cpp.js --compiler llvm --cmake true --ninja true --cppcheck true --ccache true --vcpkg true --doxygen true --gcovr true --task true
+#### Building
+FROM base AS builder
+## https://github.com/ever0de/pnpm-docker-root-bug#how-to-fix
+WORKDIR /workspace
+COPY . .
+RUN pnpm install
```
Comment (aminya) [inline dev/docker/ubuntu_20.04_node.dockerfile:35]: I prefer that we built setup-cpp outside Docker. These docker files are meant for C++ development. We can probably use `pnpm install && pnpm pack` outside Docker, and use it inside the image.

I also plan to publish these docker images on DockerHub, so we should start optimizing the docker files (like merging the RUN commands as you mentioned previously).

**Review by aminya — CHANGES_REQUESTED** [review]

File: dev/docker/ubuntu_20.04_node.dockerfile (hunk 3 above)
Comment (abeimler) [reply dev/docker/ubuntu_20.04_node.dockerfile:35]: We can use Docker for "integration testing", so maybe we don't need `ubuntu_20.04_node.dockerfile` and `ubuntu_20.04_node_legacy.dockerfile` ... and use docker for (building and) running `setup-cpp.js` + C++-builds.

So we have:
 * arch_node.dockerfile - Building and running on Arch Linux
 * ubuntu_node.dockerfile - Building and running on Ubuntu (+ nvm)
 * fefora_node.dockerfile - Building and running on Fedora
 * ubuntu.dockerfile - Installing setup-cpp (via npm) and runnung

**Comment by abeimler** [comment]
I was [experimenting](https://github.com/abeimler/setup-cpp/pull/3) with different `npm` and `node` versions + building and running `setup-cpp.js` ... 
and was wondering, do we really need `dist/node12/`, `dist/node16/` and `dist/node18/`, why not just build with the latest node version (18) and use this `setup-cpp.js` for older node version (at runtime).

Also, if I were a contributor with an older node version (12), building and testing didn't work with node 12.

**Review by aminya — APPROVED** [review]

**Pull request merged**